Add Support for SVG icons

KinanAlAttar opened the pull request:
The image is breaking for SVG for some reason, as seen below.

![Capture](https://user-images.githubusercontent.com/50109313/103315339-477a2580-49f3-11eb-9004-d120efd0164c.PNG)

Not sure why this is happening... I'm currently working on figuring this out and figured maybe you have any input or advice.

KinanAlAttar commented:
I fixed the issue, lmk if everything looks good.

johrstrom commented on apps/dashboard/app/helpers/application_helper.rb:
```diff
@@ -64,9 +68,20 @@ def app_icon_tag(app)
     if %w(fa fas far fab fal).include?(icon_uri.scheme)
       fa_icon(icon_uri.host, fa_style: icon_uri.scheme)
-    else
+    elsif !icon_uri.to_s.include? "svg"
       image_tag icon_uri.to_s, class: "app-icon", title: icon_uri.to_s, "aria-hidden": true
+    else
+      svg_tag(icon_uri.to_s, class: 'app-icon')
```
svgs need titles and aria-hidden too.  Also you probably want `if icon_uri.to_s.end_with("svg")` instead of !include.

johrstrom commented on apps/dashboard/app/apps/ood_app.rb:
```diff
@@ -181,13 +181,19 @@ def manifest_path
   end
 
-  def icon_path
+  def icon_path_png
     path.join("icon.png")
   end
```
We probably want to bury the knowledge about svgs here instead of adding 2+ interfaces. What I mean is, we want to keep the `icon_path` interface and return svg if it exists, png if it exists and nil if neither.

johrstrom replied:
actually, not nil probably `Pathname.new('')` because we're calling `file?` on the return value all over the place.

johrstrom commented:
Looks like they have a `inline_svg_tag` function that works like Rails' `image_tag`.  I'd imagine we're better of using that than calling Nokogiri. 

Maybe we do something like this. Where the helpers can then `if app.svg_icon?`.  If that's true then we can use `inline_svg_tag` API else use `image_tag`. 
```ruby
  def icon_path
    svg_icon? path.join("icon.svg") : path.join("icon.png")
  end

  def svg_icon?
    @svg_icon ||= path.join("icon.svg").file?
  end
  ```

KinanAlAttar commented:
I tried using inline_svg_tag yesterday and it was not working; it had issues with finding the svg file, even though I gave it the path

johrstrom commented:
Yea a quick check of that library and it looks they're using https://github.com/jamesmartin/inline_svg/blob/23dcb648f697833649520435d080f0e883de2744/lib/inline_svg/static_asset_finder.rb to add it to the Rails.application.assets.  

Off the top I'm not entirely sure what we'd need to do given these files are outside of the app's asset path to get `inline_svg_tag` to work, so maybe I like that Nokogiri bit a little more now.

johrstrom commented on apps/dashboard/app/helpers/application_helper.rb:
```diff
@@ -64,9 +68,20 @@ def app_icon_tag(app)
   end
 
+  def svg_tag(path, options={})
+    file = File.read(path)
+    doc = Nokogiri::HTML::DocumentFragment.parse(file)
+    svg = doc.at_css 'svg'
```
Should probably need a begin/rescue block here with some default. Also a test case to ensure the default loads when the file can't either be read or parsed (or both).

KinanAlAttar commented:
I've updated the controller method icon such that it has the proper content type for svg images (image/svg+xml instead of just image/svg); however, there seems to be an issue popping up with the QGIS icon. The QGIS app has two icons, png and svg, and I think the content type I set up does not match it. 

![qgis_icon_issue-without-svg-tag-method](https://user-images.githubusercontent.com/50109313/103375933-54f1e700-4aa9-11eb-9628-ea83c398157d.PNG)

The issue with the QGIS icon was present even with the svg tag method, but it's slightly different in that here the image was at least being rendered but not properly.  (Here I just adjusted the height and width of the icon in the developer tools to get a better idea of what was going on with the icon).

![qgis_icon_issue](https://user-images.githubusercontent.com/50109313/103376091-b619ba80-4aa9-11eb-8411-b310375999d5.PNG)

Now, I was thinking of handling cases such as QGIS, where if one icon is not rendering properly, it would then render the other. In QGIS' case it would render the png icon. Lmk if you have any better ideas.

johrstrom commented on apps/dashboard/app/apps/ood_app.rb:
```diff
@@ -181,12 +181,26 @@ def manifest_path
 
-  def icon_path
-    path.join("icon.png")
+  def svg_icon?
+    path.join("icon.svg").file?
+  end
+
+  def png_icon?
+    path.join("icon.png").file?
   end
```
Can we memorize these? seems like could be a lot of file reads. twice as many as before anyhow.

johrstrom commented on apps/dashboard/app/helpers/application_helper.rb:
```diff
@@ -48,9 +50,9 @@ def fa_icon(icon, fa_style: "fas", id: "")
 
   def app_icon_tag(app)
-    if app.icon_path.file?
+    if app.png_icon? || app.svg_icon?
```
This file seems to have a lot of updates but this is the only meaningful one. Can you clean this up a bit?

johrstrom commented:
Cool. I think the changes are good. Can we add a test case or two, at least to the model. It looks like the controller is a bit tougher to test so maybe just the model is the best effort we can do.

KinanAlAttar commented:
I feel like testing the controller would be easier if we used RSpec.

johrstrom commented:
Tests look great! I guess I'd also add `svg_icon?` and `png_icon?` interfaces to the tests. We don't need new tests, just add a assert_true (or false) statements where appropriate.

ericfranz commented on apps/dashboard/app/apps/ood_app.rb:
```diff
@@ -181,19 +181,27 @@ def manifest_path
+  
   def icon_uri
-    if icon_path_png.file? 
+    if svg_icon? || png_icon? 
```
This test is in several places. Add an `OodApp#image_icon?` method so you can do `if image_icon?` and `if app.image_icon?` instead.

johrstrom reviewed (approved):
LGTM. I think we're ready
